// used-styles/dist/es5/reactCssStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stream_1 = require("stream");
exports.process = function (chunk, line, lookupTable, callback) {
    callback(exports.getUsedStyles(chunk, lookupTable));
    return chunk;
};
var createLine = function () { return ({
    tail: '',
}); };
var classPlaceholder = 'class="';
exports.getUsedStyles = function (str, lookupTable) { return (Object.keys((str.match(/class="([^"]+)"/g) || []).slice().reduce(function (styles, className) {
    var classes = className
        .substr(classPlaceholder.length, className.length - classPlaceholder.length - 1)
        .split(' ');
    classes.forEach(function (singleClass) {
        var files = lookupTable[singleClass];
        if (files) {
            files.forEach(function (file) { return styles[file] = true; });
        }
    });
    return styles;
}, {}))); };
exports.createStyleStream = function (lookupTable, callback) {
    var line = createLine();
    var styles = {};
    var injections = [];
    var cb = function (newStyles) {
        newStyles
            .forEach(function (style) {
            if (!styles[style]) {
                styles[style] = true;
                injections.push(callback(style));
            }
        });
    };
    return new stream_1.Transform({
        transform: function (chunk, _, _callback) {
            injections = [];
            var chunkData = Buffer.from(exports.process(chunk.toString('utf-8'), line, lookupTable, cb), 'utf-8');
            _callback(undefined, injections.filter(Boolean).join('\n') + chunkData);
        },
        flush: function (cb) {
            cb();
        }
    });
};
