// used-styles/dist/es5/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var config_1 = require("./config");
var memoize_one_1 = require("./memoize-one");
exports.findLastBrace = function (data) {
    var fromIndex = 0;
    while (true) {
        var classNamePosition = data.indexOf('class=', fromIndex);
        var endBrace = data.indexOf('>', Math.max(classNamePosition, fromIndex + 1)) + 1;
        if (endBrace === 0) {
            break;
        }
        fromIndex = Math.max(classNamePosition, endBrace);
    }
    return fromIndex;
};
exports.createLine = function () { return ({
    tail: '',
}); };
function unique(data) {
    return Array.from(new Set(data));
}
exports.unique = unique;
var memoizedArray = memoize_one_1.memoizeOne(function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return args;
});
exports.getStylesInText = function (str) { return (memoizedArray.apply(void 0, config_1.isReact()
    ? exports.getStylesInReactText(str)
    : exports.getStylesInPlainText(str))); };
exports.getStylesInPlainText = function (str) { return ((str.match(/class=["']([^"]+)["']/g) || []).concat((str.match(/class=([^"'\s>]+)/g) || [])).map(function (className) { return (className.replace(/(class|'|"|=)+/g, '')); })); };
var classPlaceholder = 'class="';
var classPlaceholderLength = classPlaceholder.length;
exports.getStylesInReactText = function (str) { return ((str.match(/class="([^"]+)"/g) || []).slice().map(function (className) { return (className.substr(classPlaceholderLength, className.length - classPlaceholderLength - 1)); })); };
exports.assertIsReady = function (def) {
    if (!def.isReady) {
        throw new Error('used-styles: style definitions are not ready yet. You should `await discoverProjectStyles(...)`');
    }
};
